res/controllers/Behringer-CMDStudio4a-scripts.js

```diff
@@ -0,0 +1,257 @@
+    } else {
+        // No DEL so just cue.
+        engine.setValue(group, "cue_default", 1);
+    }
+}
+BehringerCMDStudio4a.cueRelease = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll release!
+        engine.setValue(group, "reverse", 0);
+    } else {
+        engine.setValue(group, "cue_default", 0);
+	}
+}
+
+
+// Function to deal with the scratch mode buttons.
+BehringerCMDStudio4a.scratch = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.scratchState[channel]) {
+        BehringerCMDStudio4a.scratchState[channel] = false;
+        midi.sendShortMsg(status, control, 0x00);
+    } else {
+        BehringerCMDStudio4a.scratchState[channel] = true;
+        midi.sendShortMsg(status, control, 0x01);
+    }
+}
+
+// Functions to deal with the hot-que buttons.
+BehringerCMDStudio4a.hotcuePush = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
+    if (hotcue>8) {
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    print(hotcue)
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, delete hotcue instead.
+        engine.setValue(group, "hotcue_"+hotcue+"_clear", 1);
+    } else {
+        // Del "shift" is not active, do the push hotcue function as normal.
+        engine.setValue(group, "hotcue_"+hotcue+"_activate", 1);
+    }
+}
+BehringerCMDStudio4a.hotcueRelease = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
+    if (hotcue>8) {
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, do nothing on button release.
+    } else {
+        // Del "shift" is not active, do the release hotcue function as normal.
+        engine.setValue(group, "hotcue_"+hotcue+"_activate", 0);
+    }
+}
+
+// Functions to deal with the pitch inc/dec buttons.
+BehringerCMDStudio4a.pitchDecPush = function (channel, control, value, status, group) {
+    BehringerCMDStudio4a.pitchDec[channel] = true;
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, do "key" changes instead.
+        // Check if the other button is pressed too, if so we reset the pitch.
+        if (BehringerCMDStudio4a.pitchInc[channel]) {
+            engine.setValue(group, "pitch", 0);
+        } else {
+            engine.setValue(group, "pitch_down", 1);
+        }
+    } else {
+        // DEL mode is not active, just do "rate" changes.
+        engine.setValue(group, "rate_perm_down", 1);

```

Comment:
I did originally have _temp mapping instead of _perm, but changed it for the following reasons.

1) The pitch-fader is limited to +-10% and there have been a few times when I wanted a little bit more than this. Tapping the buttons when they are mapped to _perm gives you this option when the slider is already maxed out.

2) There is no "dead-zone" option for the faders, (and on the CMD STUDIO 4a there are no "zero detents" on the faders either), so it can be *really* hard to return to zero. Using the +/- (_perm) steps instead of the fader makes this precise control easy.
Although this lack of a "zero dead-zone" was such a problem for me that I implemented an additional "zero" function by hitting both + and - together. Perhaps an adjustable "zero dead-zone" option could be added to continuous controllers in a future release??

3) "_temp" functionality is not removed by mapping to _perm. When I need a temporary increase/decrease in rate now, I just tap +/-, then tap the opposite button to cancel, i.e. I now have both _temp (albeit with an extra button tap), and _perm functionality now. This is not true the other way round, i.e. mapping to _temp means I lose all _perm stepping functionality.

If this is a blocker to acceptance, let me know and I will change it, but (IMHO) I think the current behaviour is better.